### PR TITLE
.Last.updated value after :=, closes #1885

### LINKED ISSUE #1885: Save .Last.updated after each :=
Similar to base R's `.Last.value`, data.table could save `.Last.updated` holding the number of rows updated by the last `:=`.  This could be used afterwards in production code to check that if row updates are expected then they occur and for reporting and logging.

### PR DESCRIPTION
**Pull request by jangorecki**
Closes #1885 
Now resolved: ~~but adds extra overhead of `sum(!is.na(irows))`, so should not be merged. Most likely changes needs to be pushed to C code where the actual updated number of rows is established. Not doing it now to avoid merge conflicts with rbindlist rework branch~~

### CODE REVIEW

**Comment by mattdowle** [comment]
Good idea. Would be nice to have this.
The only thing is that packages aren't allowed to write to `.GlobalEnv` without user's consent (CRAN policy).  It could be optional, default off, turned on by user. But then what if a package wished to use it inside that package?  data.table could have an object which it exports though: `.Last.updated`, a length-1 integer vector. Packages are allowed to change variables they export (like assignInNamespace does).  Then users can access the value via search() or explicitly using `data.table::.Last.updated`.
Is `.Last.updated` the best name?  Maybe it should have `nrows` or similar in the name somehow to convey it's a number of rows.

**Comment by jangorecki** [comment]
I pushed new version, using C and writing variable in data.table namespace.
The issue is that I am getting failure of one of the tests when running package check, but I am not able to reproduce it interactively. I have no clue about the reason for that.
```
Running test id 2028.14      Test 2028.14 ran without errors but failed check th
at x equals y:
> x = .Last.nrow 
First 6 of 1 (type 'integer'): [1] 1
> y = 4L 
First 6 of 1 (type 'integer'): [1] 4
Mean relative difference: 3
```

**Comment by MichaelChirico** [comment]
No manual entry?

**Comment by mattdowle** [comment]
Good catch. Could be added to `?:=` with alias at the top of that `.Rd` so `?.Last.updated` works.

**Comment by jangorecki** [comment]
I remember I made manual, but it seems I was just using `git add -u` so new files has not been added. Strangely, I must have deleted that branch locally already.

**Comment by mattdowle** [comment]
I guess this means that `R CMD check` requires exported functions to be documented but not exported variables. That's odd if so. I can't think why it shouldn't check exported variables too. Maybe we have other undocumented exported variables then.

File: R/onLoad.R [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # nocov start
 
+.Last.updated <- vector("integer", 1L) # exported variable; number of rows updated by the last := or set(), #1885
```
Comment (MichaelChirico) [inline R/onLoad.R]: This means the initial value (even if `data.table` is not loaded) is `0L`, is that preferable to `NA_integer_`?

```
library(data.table)
.Last.updated
# [1] 0
DT = data.table(a = 1)
DT[FALSE, a := 2]
.Last.updated
# [1] 0
```

Conceptually, I think `.Last.updated` is different in the two cases?

**Comment by MichaelChirico** [comment]
Is there any benefit to having a `reset_last_value` function as well, or it should be taken as immutable? I note it already seems like we can't overwrite its value manually...

**Comment by MichaelChirico** [comment]
We need to add the same note as #2837 alongside the `.Last.updated` man page, wherever it ends up:

```
DT = data.table(a = 1L)
DT[c(1, 1), a := 2:3]
DT[]
#    a
# 1: 3
.Last.updated
# [1] 2
```

**Comment by jangorecki** [comment]
I don't think we need to worry about first use, and yes, good point about duplicates on assignment, will add together with manual

**Pull request merged**